src/memcard_simulator_sm.c

```diff
@@ -0,0 +1,325 @@
+                sm_byte_counter++;
+            } else if (sm_byte_counter == 2) {
+                // LSB
+                sm_address |= data;
+                if(command_state == MC_EXECUTE_READ) {
+                    write_byte_blocking(pio, smDatWriter, MC_ACK1);
+                } else {
+                    // Otherwise send LSB
+                    write_byte_blocking(pio, smDatWriter, data);
+                }
+
+                next_state = command_state;
+                command_state = MC_IDLE;
+                sm_byte_counter = 0;
+            }
+        break;
+        case MC_EXECUTE_ID: // send mc id - used to identify which type of device this is
+            if(sm_byte_counter++ < sizeof(id_data)) {
+                write_byte_blocking(pio, smDatWriter, id_data[sm_byte_counter]);
+            } else {
+                memory_card_update_timestamp(&mc);
+                next_state = MC_IDLE;
+                in_transaction = false;
+            }
+        break;
+        case MC_EXECUTE_READ: // do a read operation
+            if(sm_byte_counter == 0) {
+                // Send ACK2
+                write_byte_blocking(pio, smDatWriter, MC_ACK2);
+                checksum = ((sm_address & 0xFF00) << 8) ^ (sm_address & 0x00FF);
+            } else if (sm_byte_counter > 0 && sm_byte_counter < 3) {
+                if(memory_card_is_sector_valid(&mc, sm_address)) {
+                    if (sm_byte_counter == 1) {
+                        // MSB
+                        write_byte_blocking(pio, smDatWriter, (sm_address & 0xFF00) >> 8);
+                    } else {
+                        // LSB
+                        write_byte_blocking(pio, smDatWriter, (sm_address & 0x00FF));
+                    }
+                } else {
+                    // Abort transaction - invalid sector
+                    write_byte_blocking(pio, smDatWriter, 0xff);
+                    next_state = MC_ABORT;
+                }
+            } else {
+                // Performing read
+                // byte counter is 3 at start here
+                uint8_t* sec_ptr = memory_card_get_sector_ptr(&mc, sm_address);
+                if ((sm_byte_counter - 3) < MC_SEC_SIZE) {
+                    write_byte_blocking(pio, smDatWriter, sec_ptr[sm_byte_counter - 3]);
+                    checksum ^= sec_ptr[sm_byte_counter - 3];
+                } else {
+                    // Send checksum
+                    write_byte_blocking(pio, smDatWriter, checksum);
+                    checksum = 0x00;
+                    next_state = MC_END;
+                }
+            }
+            sm_byte_counter++;
+        break;
+        case MC_EXECUTE_WRITE: // do a write operation
+            if(memory_card_is_sector_valid(&mc, sm_address)) {
+                uint8_t* sec_ptr = memory_card_get_sector_ptr(&mc, sm_address);
+                if(sm_byte_counter == 0) {
+                    checksum = ((sm_address & 0xFF00) << 8) ^ (sm_address & 0x00FF);
```

Comment:
```suggestion
                    checksum = ((sm_address & 0xFF00) >> 8) ^ (sm_address & 0x00FF);
```
Again shift to the right is required to isolate MSB

Reply:
Done